lib/pos_handshake_controller.go

```diff
@@ -0,0 +1,173 @@
+package lib
+
+import (
+	"github.com/decred/dcrd/lru"
+	"github.com/deso-protocol/core/bls"
+	"github.com/golang/glog"
+	"math"
+)
+
+// TODO: Replace this
+func getActiveValidators() map[bls.PublicKey]*ValidatorEntry {
+	// TODO: replace with a getter to retrieve all active validators.
+	activeValidators := []*ValidatorEntry{}
+	allValidatorsMap := make(map[bls.PublicKey]*ValidatorEntry)
+	for _, validator := range activeValidators {
+		pk := validator.VotingPublicKey
+		if pk == nil {
+			continue
+		}
+		allValidatorsMap[*pk] = validator
+	}
+
+	return allValidatorsMap
+}
+
+type HandshakeController struct {
+	rnManager              *RemoteNodeManager
+	usedNonces             lru.Cache
```

Comment:
It was an old mechanism from the previous handshake. I'm not entirely sure why we had it. There was a comment that mentioned it's to prevent self-connections, but I'm not sure whether this is a feasible threat. 

The old code also had randomized nonces rather than linearly increasing. I think linearly increasing nonces could be problematic, because if the node restarts, the nonce counter would restart as well. Maybe this could cause some handshake verack message nonce duplication.